### DISCUSSION TITLE
Cannot trigger task when app has been started programatically.

### DISCUSSION THREAD

**Discussion by anetbnd** [celery/celery#7127] [Q&A]
Hello 

I'm quite new to Celery, so maybe I have a lack of understanding.

When I have the following example implementation:
```python
import uuid
from celery import Celery

app = Celery(f'my_worker_{uuid.uuid4()}', broker="redis://localhost:6379", backend="redis")

@app.task(name='my_task')
def hello():
    return 'hello world'
```

And I start the celery app on the command line with:
```
celery -A celery_worker worker --loglevel=INFO -P gevent
```

I can trigger the task "my_task" with the following code:
```python
import uuid
from celery import Celery

app = Celery(f'my_trigger_{uuid.uuid4()}', broker="redis://localhost:6379", backend="redis")
app.signature("my_task").delay()
```

The output of celery is then:
```
[2021-11-26 17:12:03,306: INFO/MainProcess] Task my_task[015bf50b-7bb8-48ec-8c95-600cab4eb17c] received
[2021-11-26 17:12:03,309: INFO/MainProcess] Task my_task[015bf50b-7bb8-48ec-8c95-600cab4eb17c] succeeded in 0.0s: 'hello world'
```

However, if I create and start the celery app programmatically with the following code:
```python
import uuid
from celery import Celery

def create_app():
    app = Celery(f'my_worker_{uuid.uuid4()}', broker="redis://localhost:6379", backend="redis")

    @app.task(name='my_task')
    def hello():
        return 'hello world'

    return app


if __name__ == '__main__':
    app = create_app()

    app.worker_main([
        'worker',
        '--loglevel=INFO',
        '-P', 'gevent'
    ])
```

And trigger the task in the same way as above, the task is not getting executed. I just receive the following output of celery: 
```
[2021-11-26 17:14:00,506: INFO/MainProcess] Task my_task[e2993142-5454-497f-90cc-2d12f3b58315] received
```

As you can see, it successfully received the trigger for that task and was also able to find it in the celery app registry (otherwise I get an warning, saying it cannot find the task). 

I'm using python 3.9 on Windows 10 64bit with following packages:
``` 
amqp             5.0.6
billiard         3.6.4.0
celery           5.2.1
cffi             1.15.0
click            8.0.3
click-didyoumean 0.3.0
click-plugins    1.1.1
click-repl       0.2.0
colorama         0.4.4
gevent           21.8.0
greenlet         1.1.2
kombu            5.2.2
pip              21.2.3
prompt-toolkit   3.0.22
pycparser        2.21
pytz             2021.3
redis            3.5.3
setuptools       57.4.0
six              1.16.0
vine             5.0.0
wcwidth          0.2.5
zope.event       4.5.0
zope.interface   5.4.0
```

Does anyone has an idea what is wrong? Or how to solve it?

If the question appears: The reason, why I want to start the celery app programmatically and not from CLI is, that I want to be in control about the creation of the app. Maybe I want to create - for test reasons - the app with a different configuration or even mock out objects, which are used inside the task function. This would be possible by calling `create_app(default_config)` in the `__main__` module and using

```
with patch(...):
  app = create_app(test_config)
```

in a test.

**Comment by nicfix** [comment]
Hi @anetbnd, you hit a bug in gevent + celery + python3.9, that's bad luck :) 

Do you have any particular reason to use gevent insterad of the standard eventlet?

If not you can simply remove the `-P gevent` parameter when starting the worker.
```
app.worker_main([
        'worker',
        '--loglevel=INFO'
    ])
```
and it should work as expected.

For more info on the issue check out this:
https://github.com/celery/celery/issues/6627

**Reply by anetbnd** [reply]
Hi @nicfix 

Thanks for your feedback. This is really bad luck. The problem is, that eventlet is not working on Windows. See: https://github.com/celery/celery/issues/4081

Seems celery is really broken on Windows 😥

But anyway, thanks for your answer.

**Reply by anetbnd** [reply]
By the way: I tried it again with python 3.8.12 and it still does not work. 

I will retry with an linux machine.
